CDAP-4133 Ability to get the hostname of a appfabric service as a part of live-info.

sagarkapare opened the pull request:
JIRA: https://issues.cask.co/browse/CDAP-4133

live-info for the app fabric now returns the following information instead of empty object:
```java
{
    "containers": [
        {
            "container": "N/A", 
            "host": "x.y.z", 
            "instance": 0, 
            "memory": 0, 
            "name": "appfabric", 
            "type": "system_service", 
            "virtualCores": 0
        }
    ]
}
```

This allows us to locate the leader in HA setup.

sagarkapare commented:
Build is running here - http://builds.cask.co/browse/CDAP-RB304-1

chtyim commented:
Why memory is 0 and instance is 0? Also, for container, if it is not available, shouldn't it be left out / having null instead of string "N/A"?

sagarkapare commented:
Since app fabric does not run in the container I added "N/A" for the container and 0 for memory, instance and virtualCores.

Should this simply return:
 
```java
{
    "containers": [
        {
            "host": "x.y.z", 
            "name": "appfabric", 
            "type": "system_service"
        }
    ]
}
```

chtyim commented:
I think that would be better, as long as UI can handle it correctly.

sagarkapare commented:
Sounds good. I will check from UI side as well.

sagarkapare commented:
@chtyim Updated the code to not return the container related information for the app fabric service. Please take a look again. Thanks!

gsps1 commented on cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/AppFabricServiceRuntimeModule.java:
```diff
@@ -143,7 +144,7 @@ protected void configure() {
                                         .to(InMemoryMetricsServiceManager.class);
                                mapBinder.addBinding(Constants.Service.APP_FABRIC_HTTP)
-                                        .to(AppFabricServiceManager.class);
+                                        .to(InMemoryAppFabricServiceManager.class);
```
may be we can reuse this map, as standalone and in-memory both use the same map entries.

sagarkapare replied:
Moved creation of the map to the common method.

gsps1 commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java:
```diff
@@ -0,0 +1,102 @@
+
+  @Override
+  public boolean isServiceAvailable() {
+    return true;
```
so.. this is always true, as we can reach this endpoint only when app.fabric is up, as MonitorHandler runs as a part of app.fabric ?

sagarkapare replied:
correct.

gsps1 commented on cdap-proto/src/main/java/co/cask/cdap/proto/Containers.java:
```diff
@@ -46,15 +46,15 @@
 
-    public ContainerInfo(ContainerType type, String name, int instance, String container,
-                  String host, int memory, int virtualCores, Integer debugPort) {
+    public ContainerInfo(ContainerType type, String name, Integer instance, String container,
```
should add ``@Nullable``

sagarkapare replied:
Fixed.

gsps1 commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java:
```diff
@@ -0,0 +1,102 @@
+  private final InetAddress hostname;
+
+  @Inject
+  public AppFabricServiceManager(CConfiguration cConf, TwillRunnerService twillRunnerService,
```
you can implement ``MasterServiceManager`` and skip injecting unnecessary objects like ``TwillRunnerService``, etc.

sagarkapare replied:
Fixed.

gsps1 commented:
few minor comments, otherwise code changes LGTM, also make sure you are able to deserialize as expected with your new changes and also add documentation section about discovering master service in HA environment.

sagarkapare commented:
@gsps1 Addressed review comment and also verified that the deserialization of the returned JSON happens correctly. Will add doc changes as another PR. Please take a look again. Thanks!

gsps1 commented:
LGTM
